File: src/main/scala/Race/common/LineStandardForm.scala
package Race.common

object LineStandardForm {
  def apply(p1: Point, p2: Point): LineStandardForm = {
    if (p1.x == p2.x) return  vertical(p1.x)
    val (x1, y1, x2, y2) = (p1.x, p1.y, p2.x, p2.y)
    val m = (y2 - y1).toDouble / (x2 - x1).toDouble
    val a = -m
    val b = 1
    val c = a * x1 + y1
    LineStandardFormImpl(a, b, c)
  }

  def closestRealPoint(line: LineStandardForm, point: Point): (Double, Double) = {
    line match {
      case v: vertical => return (v.x, point.y)
      case _ =>
    }

    val a = line.a
    val b = line.b
    val c = -line.c
    val x = point.x.toDouble
    val y = point.y.toDouble
    val X = (((b*x - a*y) * b) - (a * c)) / (Math.pow(a, 2) + Math.pow(b, 2))
    val Y = (((-b*x + a*y) * a) - (b * c)) / (Math.pow(a, 2) + Math.pow(b, 2))
    (X, Y)
  }

  def closestPoint(line: LineStandardForm, point: Point): Point = {
    val (x, y) = closestRealPoint(line, point)
    Point(Math.round(x).toInt, Math.round(y).toInt)
  }

  case class LineStandardFormImpl(a: Double, b: Double, c: Double) extends LineStandardForm
  case class vertical(x: Int) extends LineStandardForm {
    override def a: Double = Double.NaN
    override def b: Double = Double.NaN
    override def c: Double = Double.NaN
  }
}

// line in standard form aX + bY = c
trait LineStandardForm {
  def a: Double
  def b: Double
  def c: Double
}

File: PyCar/common/line_standard_form.py
from common.point import Point
from math import pow


class LineStandardForm:
    def __init__(self, *args):
        if len(args) == 1:
            self.is_vertical = True
            self.x = float(args[0])
        if len(args) == 3:
            self.is_vertical = False
            self.a = float(args[0])
            self.b = float(args[1])
            self.c = float(args[2])

    def closest_real_point(self, p: Point) -> (float, float):
        if self.is_vertical:
            return self.x, p.y()
        (a, b, c, x, y) = self.a, self.b, -self.c, p.x(), p.y()
        _x = (((b*x - a*y) * b) - (a * c)) / (pow(a, 2) + pow(b, 2))
        _y = (((-b*x + a*y) * a) - (b * c)) / (pow(a, 2) + pow(b, 2))
        return _x, _y

    def closest_point(self, p: Point) -> Point:
        x, y = self.closest_real_point(p)
        return Point(round(x), round(y))

    @staticmethod
    def try_make(p1: Point, p2: Point):
        if p1 == p2:
            return None
        if p1.x() == p2.x():
            return LineStandardForm(p1.x())
        m = (p2.y() - p1.y()) / (p2.x() - p1.x())
        a = -m
        b = 1
        c = a * p1.x() + p1.y()
        return LineStandardForm(a, b, c)
